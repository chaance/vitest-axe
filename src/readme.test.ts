import { describe, expect, it } from "vitest";
import { configureAxe, axe } from "../src/index";
import type Axe from "axe-core";

// See https://github.com/nickcolley/jest-axe#readme
describe("readme", () => {
	describe("first readme example", () => {
		it("should demonstrate this matcher`s usage", async () => {
			let render = () => '<img src="#"/>';

			// pass anything that outputs html to axe
			let html = render();
			let results = await axe(html);
			expect(() => {
				expect(results).toHaveNoViolations();
			}).toThrowErrorMatchingSnapshot();
		});
	});

	describe("readme axe config example", () => {
		it("should demonstrate this matcher`s usage with a custom config", async () => {
			const render = () => `
          <div>
            <img src="#"/>
          </div>
        `;

			// pass anything that outputs html to axe
			const html = render();

			const results = await axe(html, {
				rules: {
					// for demonstration only, don't disable rules that need fixing.
					"image-alt": { enabled: false },
					region: { enabled: false },
				},
			});

			expect(results).toHaveNoViolations();
		});
	});

	describe("readme axe global config example", () => {
		// Global helper file (axe-helper.js)

		const configuredAxe = configureAxe({
			rules: {
				// for demonstration only, don't disable rules that need fixing.
				"image-alt": { enabled: false },
				region: { enabled: false },
			},
		});

		const exportedAxe = configuredAxe;

		// Individual test file (test.js)
		const axe = exportedAxe; // require('./axe-helper.js')

		it("should demonstrate this matcher`s usage with a default config", async () => {
			const render = () => `
          <div>
            <img src="#"/>
          </div>
        `;

			// pass anything that outputs html to axe
			const html = render();

			expect(await axe(html)).toHaveNoViolations();
		});
	});

	describe("configure custom rule", () => {
		it("should report custom rules", async () => {
			const check: any = {
				id: "demo-has-data",
				evaluate(node: Element) {
					return node.hasAttribute("data-demo-rule");
				},
				metadata: {
					impact: "serious",
					messages: {
						fail: "Error!",
					},
				},
			};
			const rule: Axe.Rule = {
				id: "demo-rule",
				selector: ".demo",
				enabled: false,
				tags: ["demo-rules"],
				any: ["demo-has-data"],
				metadata: {
					description: "Demo check",
					help: "Demo check",
					helpUrl: "https://example.com",
					tags: ["demo-rules"],
				},
			};

			const configuredAxe = configureAxe({
				globalOptions: {
					rules: [rule],
					checks: [check],
				},
				rules: {
					"demo-rule": { enabled: true },
				},
			});

			const html = `
        <html>
          <body>
            <main>
              <div class="demo">
            </main>
          </body>
        </html>
        `;

			const results = await configuredAxe(html);
			expect(results.violations[0].id).toBe("demo-rule");
		});
	});

	describe("custom configuration for user impact", () => {
		const axe = configureAxe({
			// How serious the violation is. Can be one of "minor", "moderate", "serious", or "critical".
			impactLevels: ["critical"],
		});

		it("should pass the test, because only critical violations are noted.", async () => {
			// 1 x moderate violation -> https://dequeuniversity.com/rules/axe/4.0/region?application=axeAPI
			const render = () => `
          <div>
            <div>
              <span> some content</span>
            </div>
          </div>
        `;

			// pass anything that outputs html to axe
			const html = render();

			expect(await axe(html)).toHaveNoViolations();
		});
	});
});
