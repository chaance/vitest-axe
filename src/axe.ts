import type AxeCore from "axe-core";
import { merge } from "lodash-es";
import { isHTMLElement, isHTMLString } from "./utils";
import { createRequire } from "node:module";

const require = createRequire(import.meta.url);
// eslint-disable-next-line @typescript-eslint/consistent-type-imports
const axeCore: typeof import("axe-core") = require("axe-core");
const { configure, run } = axeCore;

const AXE_RULES_COLOR = axeCore.getRules(["cat.color"]);

/**
 * Converts a HTML string or HTML element to a mounted HTML element.
 * @param html HTML element or a HTML string
 * @returns HTML element and a function to restore the document
 */
function mount(html: Element | string): [HTMLElement, () => void] {
	if (isHTMLElement(html)) {
		if (document.body.contains(html)) {
			return [html, () => undefined];
		}
		html = html.outerHTML;
	}

	if (isHTMLString(html)) {
		let originalHTML = document.body.innerHTML;
		function restore() {
			document.body.innerHTML = originalHTML;
		}

		document.body.innerHTML = html;
		return [document.body, restore];
	}

	if (typeof html === "string") {
		throw new Error(`html parameter ("${html}") has no elements`);
	}

	throw new Error(`html parameter should be an HTML string or an HTML element`);
}

/**
 * Small wrapper for axe-core#run that enables promises, default options and
 * injects html to be tested
 *
 * @param options default options to use in all instances
 *   - `globalOptions`: Global axe-core configuration (See
 *     https://github.com/dequelabs/axe-core/blob/develop/doc/API.md#api-name-axeconfigure)
 *   - Any other property will be passed as the runner configuration (See
 *     https://github.com/dequelabs/axe-core/blob/develop/doc/API.md#options-parameter)
 * @returns Instance of axe
 */
function configureAxe(
	options: AxeCore.RunOptions & {
		globalOptions?: AxeCore.Spec;
		impactLevels?: Array<AxeCore.ImpactValue>;
	} = {},
): (
	html: Element | string,
	additionalOptions?: AxeCore.RunOptions,
) => Promise<AxeCore.AxeResults> {
	let { globalOptions = {}, ...runnerOptions } = options;

	// Set the global configuration for axe-core
	// https://github.com/dequelabs/axe-core/blob/develop/doc/API.md#api-name-axeconfigure
	const { rules = [], ...otherGlobalOptions } = globalOptions;

	// Color contrast checking doesnt work in a jsdom environment. So we need to
	// identify them and disable them by default.
	const defaultRules = AXE_RULES_COLOR.map(({ ruleId: id }) => ({
		id,
		enabled: false,
	}));

	configure({
		rules: [...defaultRules, ...rules],
		...otherGlobalOptions,
	});

	/**
	 * Small wrapper for axe-core#run that enables promises, default options and
	 * injects html to be tested
	 *
	 * @param html requires a html string to be injected into the body
	 * @param additionalOptions aXe options to merge with default options
	 * @returns Promise that will resolve with axe-core#run results object
	 */
	return function axe(
		html: Element | string,
		additionalOptions: AxeCore.RunOptions = {},
	): Promise<AxeCore.AxeResults> {
		let [element, restore] = mount(html);
		let options: AxeCore.RunOptions = merge(
			{},
			runnerOptions,
			additionalOptions,
		);

		return new Promise<AxeCore.AxeResults>((resolve) => {
			run(element, options, (err, results) => {
				restore();
				if (err) throw err;
				resolve(results);
			});
		});
	};
}

const axe = configureAxe();
export { configureAxe, axe };
export type { AxeCore };
