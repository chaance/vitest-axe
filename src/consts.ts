const HORIZONTAL_LINE = "\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500"
const LINE_BREAK = "\n\n"
const MIDDLE_DOT = "\u{00B7}"

export { HORIZONTAL_LINE, LINE_BREAK, MIDDLE_DOT }
