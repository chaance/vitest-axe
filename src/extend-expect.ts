/* eslint-disable @typescript-eslint/no-unused-vars */
import type { AxeMatchers } from "./matchers";

declare global {
  namespace Vi {
    interface Assertion<T = any> extends AxeMatchers {}
    interface AsymmetricMatchersContaining extends AxeMatchers {}
  }
}
