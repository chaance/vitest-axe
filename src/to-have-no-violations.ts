import chalk from "chalk";
import type AxeCore from "axe-core";
import { printReceived, matcherHint } from "./utils";

/**
 * A custom matcher that can check aXe results for violations.
 *
 * @param results requires an instance of aXe's results object
 * (https://github.com/dequelabs/axe-core/blob/develop-2x/doc/API.md#results-object)
 * @returns Vitest matcher object
 */
export function toHaveNoViolations(
  results: AxeCore.AxeResults
): NoViolationsMatcherResult {
  if (typeof results.violations === "undefined") {
    throw new Error("No violations found in aXe results object");
  }

  let violations = filterViolations(
    results.violations,
    // TODO: Check on this error, can't find a reference to impactLevels in
    // axe-core
    results.toolOptions ? (results as any).toolOptions.impactLevels : []
  );

  function reporter(violations: AxeCore.Result[]) {
    if (violations.length === 0) {
      return [];
    }

    let lineBreak = "\n\n";
    let horizontalLine = "\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500";

    return violations
      .map((violation) => {
        let errorBody = violation.nodes
          .map((node) => {
            let selector = node.target.join(", ");
            let expectedText =
              `Expected the HTML found at $('${selector}') to have no violations:` +
              lineBreak;
            return (
              expectedText +
              chalk.grey(node.html) +
              lineBreak +
              `Received:` +
              lineBreak +
              printReceived(`${violation.help} (${violation.id})`) +
              lineBreak +
              chalk.yellow(node.failureSummary) +
              lineBreak +
              (violation.helpUrl
                ? `You can find more information on this issue here: \n${chalk.blue(
                    violation.helpUrl
                  )}`
                : "")
            );
          })
          .join(lineBreak);
        return errorBody;
      })
      .join(lineBreak + horizontalLine + lineBreak);
  }

  let formatedViolations = reporter(violations);
  let pass = formatedViolations.length === 0;

  function message(): string {
    if (pass) {
      // @ts-expect-error
      return;
    }
    return (
      // eslint-disable-next-line no-useless-concat
      matcherHint(".toHaveNoViolations") + "\n\n" + `${formatedViolations}`
    );
  }

  return { actual: violations, message, pass };
}

/**
 * Filters all violations by user impact
 * @param violations result of the accessibilty check by axe
 * @param impactLevels defines which impact level should be considered (e.g ['critical'])
 * The level of impact can be "minor", "moderate", "serious", or "critical".
 * @returns violations filtered by impact level
 */
function filterViolations(
  violations: AxeCore.Result[],
  impactLevels: Array<AxeCore.ImpactValue>
) {
  if (impactLevels && impactLevels.length > 0) {
    return violations.filter((v) => impactLevels.includes(v.impact!));
  }
  return violations;
}

interface MatcherResult {
  message(): string;
  pass: boolean;
}

export interface NoViolationsMatcherResult extends MatcherResult {
  actual: AxeCore.Result[];
}

export interface AxeMatchers {
  /**
   * A custom matcher that can check aXe results for violations.
   */
  toHaveNoViolations(): NoViolationsMatcherResult;
}
