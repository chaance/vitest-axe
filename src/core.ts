import axe from "axe-core"

export { axe as AxeCore }
