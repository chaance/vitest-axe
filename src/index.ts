import type { AxeMatchers } from "./matchers"
import { configureAxe, axe } from "./axe"
import type { AxeCore } from "./axe"

export { axe, configureAxe }
export type { AxeCore, AxeMatchers }
