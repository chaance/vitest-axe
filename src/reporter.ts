import chalk from "chalk"
import type { AxeCore } from "./core"
import { LINE_BREAK, HORIZONTAL_LINE } from "./consts"
import { printReceived } from "./utils"

const violationResolveHelpText = (violation: Pick<AxeCore.Result, "helpUrl">) => {
  return `You can find more information on this issue here: \n${chalk.blue(violation.helpUrl)}`
}

const expectedText = (node: AxeCore.NodeResult) => {
  const selector = node.target.join(", ")
  return `Expected the HTML found at $('${selector}') to have no violations:${LINE_BREAK}`
}

const reporter = (violations: AxeCore.Result[]) => {
  if (violations.length === 0) {
    return []
  }

  const delimiter = LINE_BREAK + HORIZONTAL_LINE + LINE_BREAK

  return violations
    .map((violation) => {
      const errorBody = violation.nodes
        .map((node) => {
          return (
            expectedText(node) +
            chalk.grey(node.html) +
            LINE_BREAK +
            `Received:` +
            LINE_BREAK +
            printReceived(`${violation.help} (${violation.id})`) +
            LINE_BREAK +
            chalk.yellow(node.failureSummary) +
            LINE_BREAK +
            (violation.helpUrl ? violationResolveHelpText(violation) : "")
          )
        })
        .join(LINE_BREAK)
      return errorBody
    })
    .join(delimiter)
}

export { reporter }
