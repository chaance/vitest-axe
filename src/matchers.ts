import { toHaveNoViolations } from "./to-have-no-violations"
import type { AxeMatchers, NoViolationsMatcherResult } from "./types"

export { toHaveNoViolations }
export type { NoViolationsMatcherResult, AxeMatchers }
