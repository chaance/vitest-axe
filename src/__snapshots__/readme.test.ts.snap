// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`readme > first readme example > should demonstrate this matcher\`s usage 1`] = `
[Error: expect(received).toHaveNoViolations(expected)

Expected the HTML found at $('img') to have no violations:

<img src="#">

Received:

"Images must have alternative text (image-alt)"

Fix any of the following:
  Element does not have an alt attribute
  aria-label attribute does not exist or is empty
  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
  Element has no title attribute
  Element's default semantics were not overridden with role="none" or role="presentation"

You can find more information on this issue here: 
https://dequeuniversity.com/rules/axe/4.10/image-alt?application=axeAPI

────────

Expected the HTML found at $('img') to have no violations:

<img src="#">

Received:

"All page content should be contained by landmarks (region)"

Fix any of the following:
  Some page content is not contained by landmarks

You can find more information on this issue here: 
https://dequeuniversity.com/rules/axe/4.10/region?application=axeAPI]
`;
